### PR TITLE
Causal Regression Modeling

### PR DESCRIPTION
**Pull request by psinger**
Adds new problem type causal regression modeling

Closes https://github.com/h2oai/h2o-llmstudio/issues/495

Some code is duplicated, but decided to split it into two problem types.

### CODE REVIEW

**Comment by psinger** [comment]
That's weird, import works fine for me locally, will investigate.

For docs, where would we need to add it?

**Comment by psinger** [comment]
@pascal-pfeiffer ok should be fixed, wasnt pushed due to gitignore :)

File: llm_studio/python_configs/text_causal_regression_modeling_config.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                    "BinaryCrossEntropyLoss requires num_classes == 1, "
+                    "but num_classes is set to {}.".format(self.dataset.num_classes)
+                ]
+        if self.dataset.parent_id_column not in ["None", None]:
+            errors["title"] += ["Parent ID column is not supported for classification"]
+            errors["message"] += [
+                "Parent ID column is not supported for classification datasets."
+            ]
+
+        return errors
```
Comment (pascal-pfeiffer) [inline llm_studio/python_configs/text_causal_regression_modeling_config.py:189]: there seems to be some left-over from classification

File: llm_studio/python_configs/text_causal_regression_modeling_config.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    augmentation: ConfigNLPCausalClassificationAugmentation = field(
+        default_factory=ConfigNLPCausalClassificationAugmentation
+    )
+    prediction: ConfigNLPCausalRegressionPrediction = field(
+        default_factory=ConfigNLPCausalRegressionPrediction
+    )
+    environment: ConfigNLPCausalRegressionEnvironment = field(
+        default_factory=ConfigNLPCausalRegressionEnvironment
+    )
+    logging: ConfigNLPCausalClassificationLogging = field(
```
Comment (pascal-pfeiffer) [inline llm_studio/python_configs/text_causal_regression_modeling_config.py:135]: Regression

File: llm_studio/python_configs/text_causal_regression_modeling_config.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+@dataclass
+class ConfigProblemBase(DefaultConfigProblemBase):
+    output_directory: str = f"output/{os.path.basename(__file__).split('.')[0]}"
+    experiment_name: str = field(default_factory=generate_experiment_name)
+    llm_backbone: str = "h2oai/h2o-danube2-1.8b-chat"
+
+    dataset: ConfigNLPCausalRegressionDataset = field(
+        default_factory=ConfigNLPCausalRegressionDataset
+    )
+    tokenizer: ConfigNLPCausalClassificationTokenizer = field(
```
Comment (pascal-pfeiffer) [inline llm_studio/python_configs/text_causal_regression_modeling_config.py:117]: Regression

File: llm_studio/python_configs/text_causal_regression_modeling_config.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        self._visibility["num_classes"] = -1
+
+
+@dataclass
+class ConfigNLPCausalRegressionTraining(ConfigNLPCausalClassificationTraining):
+    loss_class: Any = text_causal_regression_modeling_losses.Losses
+    loss_function: str = "MSELoss"
+
+    learning_rate: float = 0.0001
+    differential_learning_rate_layers: Tuple[str, ...] = ("classification_head",)
```
Comment (pascal-pfeiffer) [inline llm_studio/python_configs/text_causal_regression_modeling_config.py:51]: regression_head

File: llm_studio/python_configs/text_causal_regression_modeling_config.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    differential_learning_rate: float = 0.00001
+
+    def __post_init__(self):
+        super().__post_init__()
+        self._possible_values["loss_function"] = self.loss_class.names()
+
+        self._possible_values["differential_learning_rate_layers"] = (
+            possible_values.String(
+                values=("backbone", "embed", "classification_head"),
```
Comment (pascal-pfeiffer) [inline llm_studio/python_configs/text_causal_regression_modeling_config.py:60]: regression_head

File: llm_studio/src/metrics/text_causal_regression_modeling_metrics.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+def mae_score(
+    cfg: Any,
+    results: Dict,
+    val_df: pd.DataFrame,
+    raw_results: bool = False,
+) -> Union[NDArray, Tuple[NDArray, List[str]]]:
+    predicted_text = np.array([float(text) for text in results["predicted_text"]])
+    target_text = np.array([float(text) for text in results["target_text"]])
+    return (target_text - predicted_text).astype("float")
```
Comment (pascal-pfeiffer) [inline llm_studio/src/metrics/text_causal_regression_modeling_metrics.py:30]: isn't there an `abs` missing?

**Review by pascal-pfeiffer — CHANGES_REQUESTED** [review]
Working quite nicely, thank you!

Indeed, more documentation changes are probably not needed. Some suggestions/comments/questions in the code below.

**Review by pascal-pfeiffer — APPROVED** [review]
Thank you! lgtm

**Pull request merged**